hutch_python/tests/test_ipython_session_timer.py

```diff
@@ -0,0 +1,52 @@
+import time
+import unittest
+
+import pytest
+from test_ipython_log import FakeIPython
+
+from hutch_python.ipython_session_timer import IPythonSessionTimer
+
+
+@pytest.fixture(scope='function')
+def fake_ipython():
+    fake_ipython = FakeIPython()
+    return fake_ipython
+
+
+@pytest.fixture(scope='function')
+def session_timer(fake_ipython):
+    session_timer = IPythonSessionTimer(fake_ipython)
+    session_timer.max_idle_time = 20.0
+    session_timer.user_active = True
+    session_timer.curr_time = time.monotonic()
+    session_timer.last_active_time = session_timer.curr_time - 10.0
+    session_timer.idle_time = session_timer.curr_time - session_timer.last_active_time
+    return session_timer
+
+
+def test_set_user_active(session_timer):
+    assert session_timer.user_active
+
+
+def test_set_user_inactive(session_timer):
+    session_timer.user_active = False
+    assert not session_timer.user_active
+
+
+def test_set_idle_time(session_timer):
+    assert session_timer.idle_time == 10.0
+
+
+# Skipping tests for _start_session() where self.user_active==True because this enters an
+# infinite while loop.
```

Comment:
It's probably worthwhile finding a way to test and make sure that a session with `self.user_active` isn't exited. This might look something like running `_start_session` in a `threading.Thread` as done in `load_ipython_extension`. The infinite while loop only applies when this runs in the foreground thread. We could then set the max idle time to e.g. 1s and verify that ask_exit and exit never get called after 5s or so. Then, in the same test, we could flip `self.user_active` back to `False` and see that `ask_exit`, etc. get called soon after.

Reply:
This might take some time to set up. Could we merge without this one?

Reply:
Yes, I think it's healthy to merge and consider this later (if ever)